mythril/src/acpi/rsdt.rs

```diff
@@ -229,3 +234,157 @@ impl<'a> Iterator for RSDTIterator<'a> {
+    buffer.extend(table);
+
+    // The SDT length value is the value of the entire SDT including
+    // the header.
+    let sdt_len = offsets::CREATOR_REVISION.end + table.len();
+    if buffer.len() < sdt_len {
+        return Err(Error::InvalidValue(format!(
+            "Buffer length should be at least `{}` but was `{}`",
+            sdt_len,
+            buffer.len()
+        )));
+    }
+    // Fill in the SDT header with the implementations values
+    buffer[offsets::SIGNATURE].copy_from_slice(signature);
+    NativeEndian::write_u32(&mut buffer[offsets::LENGTH], sdt_len as u32);
+    buffer[offsets::REVISION] = revision;
+    NativeEndian::write_u32(
+        &mut buffer[offsets::CREATOR_REVISION],
+        creator_revision,
+    );
+
+    // After the SDT header and body have been populated calculate the
+    // table checksum.
+    buffer[offsets::CHECKSUM] = calc_checksum(&buffer);
+
+    Ok(buffer)
+}
+
+/// Builder trait for System Descriptor Tables
+pub trait SDTBuilder {
+    /// The identifying signature for this table type.
+    const SIGNATURE: [u8; 4];
+
+    /// The revision of the tabe
+    fn revision(&self) -> u8 {
+        0
+    }
+
+    /// The revision of the tabe
+    fn creator_revision(&self) -> u32 {
+        0
+    }
+
+    // TODO(dlrobertson): Can we rely on encoding the SDT table
+    // without access to the guest address space? Do we need to
+    // know the address of anything we place in the guest?
+
+    /// Attempt to encode the SDT table
+    fn encode_table(&self) -> Result<Vec<u8>>;
+
+    /// Encode the entire SDT including the header and body based on the
+    /// implemented methods.
+    fn encode_sdt(&self) -> Result<Vec<u8>> {
+        build_sdt(
+            &Self::SIGNATURE,
+            self.revision(),
+            self.creator_revision(),
+            &self.encode_table()?,
+        )
+    }
+}
+
+/// Builder structure for the RSDT
+pub(super) struct RSDTBuilder {
+    map: BTreeMap<[u8; 4], Vec<u8>>,
+}
+
+impl RSDTBuilder {
+    /// Create a new XSDT Builder.
+    pub(super) fn new() -> RSDTBuilder {
+        RSDTBuilder {
+            map: BTreeMap::new(),
+        }
+    }
+
+    /// Add the given System Descriptor Table to the RSDT.
+    pub(super) fn add_sdt<T: SDTBuilder>(&mut self, builder: T) -> Result<()> {
+        if !self.map.contains_key(&T::SIGNATURE) {
+            self.map.insert(T::SIGNATURE, builder.encode_sdt()?);
+            Ok(())
+        } else {
+            Err(Error::InvalidValue(format!(
+                "The key `{}` already exists",
+                str::from_utf8(&T::SIGNATURE).unwrap()
+            )))
+        }
+    }
+
+    /// Create the encoded ACPI table.
+    pub fn build(
+        &self,
+        fw_cfg_builder: &mut QemuFwCfgBuilder,
+        table_loader: &mut TableLoaderBuilder,
+    ) -> Result<()> {
+        let table_length = self.map.iter().count() * 8;
+        let xsdt = build_sdt(b"XSDT", 0, 0, &vec![0x00; table_length])?;
```

Comment:
We don't know how big this buffer will be, and we don't know how big the entire table loader buffer will be, so I'm not sure the best way to move away from alloc is, but there are some options.